File: sdk/js/src/auth-store.ts
export type AuthModel = Record<string, unknown> | null;

type Listener = (token: string, model: AuthModel) => void;

/** Holds the current auth token + record/admin, persisted to
 * `localStorage` when available (browser) so a page reload doesn't log the
 * user out. In non-browser environments (SSR, React Native without a
 * storage polyfill, CLI scripts) it just holds state in memory for the
 * process lifetime — set your own `save`/`clear` if you need custom
 * persistence (e.g. `AsyncStorage` on React Native, an httpOnly cookie on
 * the server). */
export class AuthStore {
  private storageKey: string;
  private listeners: Listener[] = [];
  private _token = "";
  private _model: AuthModel = null;

  constructor(storageKey = "cratebase_auth") {
    this.storageKey = storageKey;
    this.loadFromStorage();
  }

  get token(): string {
    return this._token;
  }

  get model(): AuthModel {
    return this._model;
  }

  get isValid(): boolean {
    if (!this._token) return false;
    const payload = decodeJwtPayload(this._token);
    if (!payload || typeof payload.exp !== "number") return true;
    return payload.exp * 1000 > Date.now();
  }

  save(token: string, model: AuthModel): void {
    this._token = token;
    this._model = model;
    this.persist();
    this.emit();
  }

  clear(): void {
    this._token = "";
    this._model = null;
    this.persist();
    this.emit();
  }

  onChange(listener: Listener): () => void {
    this.listeners.push(listener);
    return () => {
      this.listeners = this.listeners.filter((l) => l !== listener);
    };
  }

  private emit(): void {
    for (const listener of this.listeners) listener(this._token, this._model);
  }

  private persist(): void {
    const storage = getLocalStorage();
    if (!storage) return;
    if (this._token) {
      storage.setItem(this.storageKey, JSON.stringify({ token: this._token, model: this._model }));
    } else {
      storage.removeItem(this.storageKey);
    }
  }

  private loadFromStorage(): void {
    const storage = getLocalStorage();
    if (!storage) return;
    const raw = storage.getItem(this.storageKey);
    if (!raw) return;
    try {
      const parsed = JSON.parse(raw) as { token: string; model: AuthModel };
      this._token = parsed.token ?? "";
      this._model = parsed.model ?? null;
    } catch {
      // corrupt storage entry: ignore, start logged out
    }
  }
}

function getLocalStorage(): Storage | null {
  try {
    return typeof window !== "undefined" ? window.localStorage : null;
  } catch {
    return null;
  }
}

function decodeJwtPayload(token: string): { exp?: number } | null {
  const parts = token.split(".");
  if (parts.length !== 3) return null;
  try {
    const json = typeof atob === "function" ? atob(parts[1]) : Buffer.from(parts[1], "base64").toString("utf-8");
    return JSON.parse(json);
  } catch {
    return null;
  }
}

File: sdk/python/src/cratebase/auth_store.py
"""Holds the current auth token + record/admin.

Mirrors `sdk/js/src/auth-store.ts`, but Python has no browser
`localStorage` to persist into automatically, so `AuthStore` simply holds
state in memory for the process lifetime -- matching the JS SDK's own
non-browser fallback (SSR, React Native without a storage polyfill, CLI
scripts). Attach an `on_change` listener, or subclass and override
`save`/`clear`, if you need custom persistence (a file, a keyring entry, an
httpOnly cookie, a session table).
"""

from __future__ import annotations

import base64
import binascii
import json
import time
from typing import Any, Callable, Dict, List, Optional

AuthModel = Optional[Dict[str, Any]]
Listener = Callable[[str, AuthModel], None]


class AuthStore:
    def __init__(self) -> None:
        self._token: str = ""
        self._model: AuthModel = None
        self._listeners: List[Listener] = []

    @property
    def token(self) -> str:
        return self._token

    @property
    def model(self) -> AuthModel:
        return self._model

    @property
    def is_valid(self) -> bool:
        """`True` if a token is present and (when it's a JWT with an `exp`
        claim) not yet expired. Tokens without a parseable `exp` are
        assumed valid -- the server is still the source of truth."""
        if not self._token:
            return False
        payload = _decode_jwt_payload(self._token)
        if payload is None or not isinstance(payload.get("exp"), (int, float)):
            return True
        return payload["exp"] * 1000 > time.time() * 1000

    def save(self, token: str, model: AuthModel) -> None:
        self._token = token
        self._model = model
        self._emit()

    def clear(self) -> None:
        self._token = ""
        self._model = None
        self._emit()

    def on_change(self, listener: Listener) -> Callable[[], None]:
        """Registers `listener` to be called with `(token, model)` on every
        `save`/`clear`. Returns an unsubscribe function."""
        self._listeners.append(listener)

        def unsubscribe() -> None:
            if listener in self._listeners:
                self._listeners.remove(listener)

        return unsubscribe

    def _emit(self) -> None:
        for listener in list(self._listeners):
            listener(self._token, self._model)


def _decode_jwt_payload(token: str) -> Optional[Dict[str, Any]]:
    parts = token.split(".")
    if len(parts) != 3:
        return None
    try:
        padded = parts[1] + "=" * (-len(parts[1]) % 4)
        decoded = base64.urlsafe_b64decode(padded)
        return json.loads(decoded)
    except (ValueError, binascii.Error, json.JSONDecodeError):
        return None
